File: GameServer/TypeScript/Game/new-svelte/src/lib/battle/battle-attributes.ts
import { IBattlerAttribute, EAttribute } from "$lib/api";
import { normalizeText } from "$lib/common";

export class BattleAttributes {
    static #attributesMaxId = Object.values(EAttribute)[Object.values(EAttribute).length - 1] as number;
    attributes: number[];

    constructor(attList: IBattlerAttribute[], calcDerivedStats: boolean = true) {
        this.attributes = new Array<number>(BattleAttributes.#attributesMaxId);
        this.attributes.fill(0, 0, BattleAttributes.#attributesMaxId);
        attList.forEach(att => this.attributes[att.attributeId] += att.amount);
        if (calcDerivedStats) {
            this.#calculateDerivedStats()
        }
    }

    getValue(attId: number) {
        return this.attributes[attId];
    }

    getAttributeMap(includeZeroes: boolean = false) {
        return this.attributes
            .map((att, i) => ({ name: normalizeText(EAttribute[i]), value: att }))
            .filter(att => att.value != 0 || includeZeroes);
    }

    #calculateDerivedStats() {
        this.attributes[EAttribute.MaxHealth] += 50 + (20 * this.attributes[EAttribute.Endurance]) + (5 * this.attributes[EAttribute.Strength]);
        this.attributes[EAttribute.Defense] += 2 + this.attributes[EAttribute.Endurance] + (0.5 * this.attributes[EAttribute.Agility]);
        this.attributes[EAttribute.CooldownRecovery] += (0.4 * this.attributes[EAttribute.Agility]) + (0.1 * this.attributes[EAttribute.Dexterity]);
        this.attributes[EAttribute.DropBonus] += Math.log10(this.attributes[EAttribute.Luck]);
        // this.attributes[EAttribute.CriticalChance] = 0;
        // this.attributes[EAttribute.CriticalDamage] = 0;
        // this.attributes[EAttribute.DodgeChance] = 0;
        // this.attributes[EAttribute.BlockChance] = 0;
        // this.attributes[EAttribute.BlockReduction] = 0;
    }
}

File: Game.Core/BattleSimulation/BattleAttributes.cs
using static Game.Core.EAttribute;

namespace Game.Core.BattleSimulation
{
    /// <summary>
    /// Represents an aggregated collection of <see cref="BattlerAttribute"/>s.
    /// </summary>
    public class BattleAttributes
    {
        private static readonly int _attributesMaxId = GetMaxAttribute();
        private readonly List<double> _attributes = GetEmptyAttributeList();

        /// <summary>
        /// Gets the associated value of an attribute by its <see cref="EAttribute"/> representation.
        /// </summary>
        /// <param name="index"></param>
        /// <returns>The value of the attribute.</returns>
        public double this[EAttribute index]
        {
            get => _attributes[(int)index];
            set => _attributes[(int)index] = value;
        }

        /// <summary>
        /// Creates a new aggregate from an <see cref="IEnumerable{T}"/> of <see cref="BattlerAttribute"/>.
        /// </summary>
        /// <param name="attributes"></param>
        public BattleAttributes(IEnumerable<BattlerAttribute> attributes)
        {
            foreach (var att in attributes)
            {
                this[att.AttributeId] += (double)att.Amount;
            }

            CalculateDerivedValues();
        }

        private static List<double> GetEmptyAttributeList()
        {
            return Enumerable.Repeat(0.0, _attributesMaxId + 1).ToList();
        }

        private static int GetMaxAttribute()
        {
            var attributeValues = Enum.GetValues<EAttribute>();
            return (int)attributeValues.Max();
        }

        private void CalculateDerivedValues()
        {
            this[MaxHealth] += 50.0 + (20.0 * this[Endurance]) + (5 * this[Strength]);
            this[Defense] += 2.0 + this[Endurance] + (0.5 * this[Agility]);
            this[CooldownRecovery] += (0.4 * this[Agility]) + (0.1 * this[Dexterity]);
            this[DropBonus] += this[Luck] > 0.0 ? Math.Log10(this[Luck]) : 0.0;
            //this[CriticalChance] = 0.0;
            //this[CriticalDamage] = 0.0;
            //this[DodgeChance] = 0.0;
            //this[BlockChance] = 0.0;
            //this[BlockReduction] = 0.0;
        }
    }
}
